### ISSUE TITLE
How do I train a model using my own data

### ISSUE THREAD

**Issue by philip-pan** [torrvision/crfasrnn#17]
I use the NYU V2 dataset as my traing set. The labels are created as follows. I first create a picture each pixel corresponding to its label and create labels_lmdb using these pictures by convert_imageset in the dictory caffe-root/tools. images_lmdb is created using images in the dataset by that tool convert_imageset directly.

I have trained my model using Tesla K20m, but the loss of the nueral network seems oscillating all the time after 400 iterations. The whole process of 400 iterations takes about 10 hours. I think that there is something wrong with my configurations. I am wondering whether you could help me to find the problems.

I modified the file "TVG_CRFRNN_COCO_VOC.prototxt" provided by the project of CRFasRNN in the dicrectory crfasrnn/python-scripts. Details as follows. 

#I added the data layer at the begining of that prototxt file to replace the original input part (line 3-8):
layers {
  name: "data"
  type: DATA
  top: "data"
  include {
    phase: TRAIN
  }
  transform_param {
    mean_value: 130.4265
    mean_value: 111.4584
    mean_value: 103.3727
  }
  data_param {
    source: "./indoor_images_lmdb"
    batch_size: 1
    backend: LMDB
  }
}
layers {
  name: "label"
  type: DATA
  top: "label"
  include {
    phase: TRAIN
  }
  data_param {
    source: "./indoor_labels_lmdb"
    batch_size: 1
    backend: LMDB
  }
}

#At the end of that file, I added a SOFTMAX_LOSS layer:
layers { type: SOFTMAX_LOSS name: 'loss' top: 'loss'
  bottom: 'pred' bottom: 'label'
  loss_param { normalize: false }
  }

#The file "solver.prototxt" was created as follows:
net: "CRFasRNN_train.prototxt"
test_iter: 500
test_interval: 100000
display: 250
lr_policy: "fixed"
base_lr: 1e-13
momentum: 0.99
weight_decay:0.0016
max_iter: 10000
snapshot: 1000
snapshot_prefix: "train"
test_initialization: false
solver_mode: GPU

#The file "solve.py" was created as follows:
caffe_root = '../caffe-crfrnn/'
import sys
sys.path.insert(0, caffe_root + 'python')
import caffe
import numpy as np
base_weights = 'TVG_CRFRNN_COCO_VOC.caffemodel'
solver = caffe.SGDSolver('solver.prototxt')
solver.net.copy_from(base_weights)
solver.step(10000)  
The file "TVG_CRFRNN_COCO_VOC.caffemodel" is also provided in the directory crfasrnn/python-script.

**Comment by bittnt** [comment]
Hey, @philip-pan which loss do you mean? For segmentation, after a few interactions, you should see the iOU score increase. Loss output for one instance ( one instance per batch) does not provide useful information. Hope this helps u.

**Comment by mehtasachin** [comment]
Hi,

I am trying to train my data, but I am not able to do so. I am getting below error:

I0101 11:24:10.820197  5307 net.cpp:105] Top shape: 1 3 360 480 (518400)
I0101 11:24:10.820204  5307 layer_factory.hpp:78] Creating layer label
I0101 11:24:10.820214  5307 net.cpp:69] Creating Layer label
F0101 11:24:10.820222  5307 net.cpp:354] Duplicate blobs produced by multiple sources.
*** Check failure stack trace: ***
    @     0x7f49b5cbadaa  (unknown)
    @     0x7f49b5cbace4  (unknown)
    @     0x7f49b5cba6e6  (unknown)
    @     0x7f49b5cbd687  (unknown)
    @     0x7f49b5fa3e82  caffe::Net<>::AppendTop()
    @     0x7f49b5fab29d  caffe::Net<>::Init()
    @     0x7f49b5face2e  caffe::Net<>::Net()
    @     0x7f49b6073000  caffe::Solver<>::InitTrainNet()
    @     0x7f49b60740d3  caffe::Solver<>::Init()
    @     0x7f49b60742a6  caffe::Solver<>::Solver()
    @           0x40b570  caffe::GetSolver<>()
    @           0x406c24  train()
    @           0x404a61  main
    @     0x7f49b51ccec5  (unknown)
    @           0x405676  (unknown)
    @              (nil)  (unknown)

Could you please provide me some insight about the above mentioned error? I've attached my solver and training files. Also, I am creating the lmdb files as:
image_loc 0

Here, 0 should be a class label but I am not sure how to provide it as we are providing the labels as another image.
[solver.txt](https://github.com/torrvision/crfasrnn/files/76006/solver.txt)
[training.txt](https://github.com/torrvision/crfasrnn/files/76007/training.txt)

**Comment by bittnt** [comment]
Looks like your the first few lines in training.txt contain mistakes. Why you have multiple outputs called label? This might be the reason why you get that error messages. Check FCN pascal context training script. 

Also you would better set test data and label and return accuracy. Otherwise, it is impossible to know when you should stop the training.

**Comment by thuanvh** [comment]
Hi @bittnt .
I see a log line 

I0105 11:59:55.087455  4092 multi_stage_meanfield.cpp:45] This implementation has not been tested batch size > 1.

What is the value batch_size you used in training?

**Comment by bittnt** [comment]
@thuanvh batch_size=1.

**Comment by jay98** [comment]
@thuanvh Were you able to fix the error as I have a similar error. I have the batch size as 1 but it's still giving me the same error.

Thanks 
-Jay

**Comment by ThienAnh** [comment]
@bittnt  Can i training with number of class more than 21 (ex: I want to train with 40 class)?
